File: android/app/src/main/kotlin/dev/crossflow/android/Protocol.kt
package dev.crossflow.android

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import kotlinx.serialization.decodeFromString

@Serializable
data class ClipMessage(
    val type: String = "clipboard",
    val content: String,
    val source: String
)

object Protocol {
    const val PORT = 35647
    const val SERVICE_TYPE = "_crossflow._tcp."
    private val json = Json { ignoreUnknownKeys = true }

    fun encode(msg: ClipMessage): String = json.encodeToString(msg) + "\n"

    fun decode(line: String): ClipMessage? = try {
        json.decodeFromString<ClipMessage>(line.trim())
    } catch (e: Exception) { null }
}

File: macos/Sources/Protocol.swift
import Foundation

// Shared message format — identical across all platforms
struct ClipMessage: Codable {
    var type: String = "clipboard"
    var content: String
    var source: String
}

enum Protocol {
    static let port: UInt16 = 35647
    static let serviceType  = "_crossflow._tcp."

    static func encode(_ msg: ClipMessage) -> Data {
        let line = (try? JSONEncoder().encode(msg)).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"
        return (line + "\n").data(using: .utf8) ?? Data()
    }

    static func decode(_ data: Data) -> ClipMessage? {
        guard let str = String(data: data, encoding: .utf8)?
                .trimmingCharacters(in: .whitespacesAndNewlines),
              !str.isEmpty else { return nil }
        return try? JSONDecoder().decode(ClipMessage.self, from: Data(str.utf8))
    }
}
